mirrord/agent/src/outgoing.rs

```diff
@@ -160,89 +78,266 @@ impl TcpOutgoingApi {
-                            warn!("interceptor_task -> read connection_id {:#?} failed with {:#?}", connection_id, err);
-                            writers.remove(&connection_id);
-
-                            let daemon_message = DaemonTcpOutgoing::Close(connection_id);
-                            daemon_tx.send(daemon_message).await?
-                        }
-                        None => {
-                            trace!("interceptor_task -> close connection {:#?}", connection_id);
-                            writers.remove(&connection_id);
+    #[tracing::instrument(
+        level = "trace",
+        skip(read),
+        fields(read = ?read.as_ref().map(|res| res.as_ref().map(|bytes| bytes.len()))),
+        ret,
+        err(Debug)
+    )]
+    async fn handle_connection_read(
+        &mut self,
+        connection_id: ConnectionId,
+        read: Option<io::Result<Bytes>>,
+    ) -> Result<(), SendError<DaemonTcpOutgoing>> {
+        match read {
+            // TODO(alex): PROTOCOL
+            // We shouldn't return it as a `Result<T, ResponseError>` since we lose track of
+            // connection id and it doesn't really make sense to do it, but we don't
+            // want to break the protocol so we'll still wrap it with Ok() and if we
+            // error we just close the connection.
+            Some(Ok(read)) => {
+                let message = DaemonTcpOutgoing::Read(Ok(DaemonRead {
+                    connection_id,
+                    bytes: read.to_vec(),
+                }));
 
-                            let daemon_message = DaemonTcpOutgoing::Close(connection_id);
-                            daemon_tx.send(daemon_message).await?
-                        }
-                    }
-                }
-                else => {
-                    // We have no more data coming from any of the remote hosts.
-                    warn!("interceptor_task -> no messages left");
-                    break;
+                self.daemon_tx.send(message).await?;
+            }
+
+            Some(Err(error)) => {
+                tracing::trace!(
+                    ?error,
+                    connection_id,
+                    "Reading from peer connection failed, sending close message.",
+                );
+
+                self.writers.remove(&connection_id);
+                let daemon_message = DaemonTcpOutgoing::Close(connection_id);
+                self.daemon_tx.send(daemon_message).await?;
+            }
+
+            None => {
+                tracing::trace!(
+                    connection_id,
+                    "Peer connection shutdown, sending 0-sized read message.",
+                );
+
+                let daemon_message = DaemonTcpOutgoing::Read(Ok(DaemonRead {
+                    connection_id,
+                    bytes: vec![],
+                }));
+
+                self.daemon_tx.send(daemon_message).await?;
+
+                if !self.writers.contains_key(&connection_id) {
```

Comment:
sorry if you told me that already, but why can't we remove the reader of this connection in this case?
If we're here it means the user's application shut down from its side, and the connection peer has shut down from its side, right?